### PR TITLE
REGR: Different results from DataFrame.apply and str accessor

### LINKED ISSUE #38979: BUG: Different results from DataFrame.apply and str accessor
- [x] I have checked that this issue has not already been reported.

- [x] I have confirmed this bug exists on the latest version of pandas.

- [x] (optional) I have confirmed this bug exists on the master branch of pandas.

---

**Note**: Please read [this guide](https://matthewrocklin.com/blog/work/2018/02/28/minimal-bug-reports) detailing how to provide the necessary information for us to reproduce your bug.

#### Code Sample, a copy-pastable example

```python
# Your code here
import pandas as pd

print(pd.__version__)

df = pd.DataFrame(zip('abc', 'def'))
print(df.apply(lambda f: "/".join(f), axis=1))
print(df.apply(lambda f: "/".join(f.str.upper()), axis=1))
```

#### Problem description
Using the `str` accessor gives different results on 1.1.5 and 1.2.x:
```
1.1.5
0    a/d
1    b/e
2    c/f
dtype: object
0    A/D
1    B/E
2    C/F
dtype: object

1.2.0+22.g6cdb4e7fb
0    a/d
1    b/e
2    c/f
dtype: object
0    A/D
1    A/D
2    A/D
dtype: object

```

#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit           : 6cdb4e7fbf4ba9a8fc83af550d866765f64ecdd9
python           : 3.7.9.final.0
python-bits      : 64
OS               : Windows
OS-release       : 10
Version          : 10.0.19041
machine          : AMD64
processor        : Intel64 Family 6 Model 63 Stepping 2, GenuineIntel
byteorder        : little
LC_ALL           : None
LANG             : None
LOCALE           : None.None

pandas           : 1.2.0+22.g6cdb4e7fb
numpy            : 1.19.5
pytz             : 2020.5
dateutil         : 2.8.1
pip              : 20.3.3
setuptools       : 51.0.0.post20201207
Cython           : 0.29.21
pytest           : None
hypothesis       : None
sphinx           : None
blosc            : None
feather          : None
xlsxwriter       : None
lxml.etree       : None
html5lib         : None
pymysql          : None
psycopg2         : None
jinja2           : None
IPython          : None
pandas_datareader: None
bs4              : None
bottleneck       : None
fsspec           : None
fastparquet      : None
gcsfs            : None
matplotlib       : None
numexpr          : None
odfpy            : None
openpyxl         : None
pandas_gbq       : None
pyarrow          : None
pyxlsb           : None
s3fs             : None
scipy            : None
sqlalchemy       : None
tables           : None
tabulate         : None
xarray           : None
xlrd             : None
xlwt             : None
numba            : None

</details>

### PR DESCRIPTION
**Pull request by simonjayhawkins**
- [ ] closes #38979
- [ ] tests added / passed
- [ ] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing.html#code-standards) for how to run them
- [ ] whatsnew entry

@jbrockmendel despite my comment https://github.com/pandas-dev/pandas/issues/38979#issuecomment-760470446, I had no joy trying to locate the caching issue.

### CODE REVIEW

File: pandas/tests/test_strings.py [hunk 1]
```diff
@@ -3670,3 +3670,11 @@ def test_str_get_stringarray_multiple_nans():
     result = s.str.get(2)
     expected = Series(pd.array([pd.NA, pd.NA, pd.NA, "c"]))
     tm.assert_series_equal(result, expected)
+
+
+def test_str_accessor_in_apply_func():
+    # https://github.com/pandas-dev/pandas/issues/38979
+    df = pd.DataFrame(zip("abc", "def"))
+    expected = df.apply(lambda f: "/".join(f), axis=1).str.upper()
```
Comment (jbrockmendel) [inline pandas/tests/test_strings.py:3678]: might be worth writing this out explicitly?

Comment (simonjayhawkins) [reply pandas/tests/test_strings.py]: good call.

**Comment by jreback** [comment]
thanks @simonjayhawkins

**Comment by jreback** [comment]
@meeseeksdev backport 1.2.x

**Pull request merged**